### PR TITLE
Download: don't create empty cache file when none existed before

### LINKED ISSUE #123: ccm-fetch fails to fetch the profile if the cache is empty
EDG::WP4::CCM::Fetch::Download::download() uses the timestamp of the last downloaded profile, which is cached under /var/lib/ccm/data, to decide if the profile has been changed on the server side. However, if there is no cached profile, then the code creates an empty file just before checking the timestamp:

    if (!-f $cache) {
        CAF::FileWriter->new($cache, log => $self)->close();
    }

    my @st = stat($cache) or die "Unable to stat profile cache: $cache ($!)";

meaning the completely wrong timestamp is sent to the server, and the profile will not be downloaded until it is forcibly recompiled.

I cannot see how an empty file could be useful here, so the code should be changed never to create an empty file, and not to pass the timestamp to ->retrieve() if there was no cache file. Adding the If-Modified-Since header in ->retrieve() should then be made conditional on $time being defined.

### PR DESCRIPTION
**Pull request by stdweird**
Fixes #123

### CODE REVIEW

**Comment by jrha** [comment]
Typo in commit message (previou)

File: src/main/perl/Fetch/Download.pm [hunk 1]
```diff
@@ -94,17 +94,29 @@ sub retrieve
     my $ua = LWP::UserAgent->new();
     my $rq = HTTP::Request->new(GET => $url);
 
-    my $ht = scalar(localtime($time));
-    if (!$self->{FORCE}) {
+    # human readable time (if not defined, a 304 is an error)
```
Comment (jrha) [inline src/main/perl/Fetch/Download.pm]: I'm not sure what this comment is trying to say.

If human readable time is not defined, then treat as a 304?

Comment (stdweird) [inline src/main/perl/Fetch/Download.pm]: i'll replace with `If human readable time is not defined, then treat a 304 repsonse as an error` (see code below)

File: src/main/perl/Fetch/Download.pm [hunk 2]
```diff
@@ -94,17 +94,29 @@ sub retrieve
...
     $rq->header("Accept-Encoding" => join(" ", qw(gzip x-gzip x-bzip2 deflate)));
     my $rs = $ua->request($rq);
     if ($rs->code() == 304) {
-        $self->verbose("No changes on $url since $ht");
-        return 0;
+        if (defined($ht)) {
+            $self->verbose("No changes on $url since $ht");
+            return 0;
+        } else {
+            $self->warn("Unexpected no changes on $url without if_modified_since set in request");
```
Comment (ned21) [inline src/main/perl/Fetch/Download.pm]: Nitpick: but this doesn't match the comment of line 97. If we get back a 304 then we get no data so we can't continue, so I think this is an error and the comment is correct. I think the message should be:
```
Server responded with code 304 for $url even though if_modified_since was not set in request. No profile retrieved.
```

**Comment by ned21** [comment]
One minor remark, otherwise LGTM. Thanks for quick fix.

**Comment by stdweird** [comment]
@ned21 remark fixed, i also report it with error now. but CCM is a bit a mess with inconsistent reporting (`warn`, `error` and `die` are used).

**Pull request merged**